File: backend/app/schemas/bills.py
from __future__ import annotations

from datetime import date
from typing import Any

from pydantic import BaseModel

from app.schemas.common import AnalysisState, DataGap
from app.schemas.votes import VoteListItem


class BillDeathInfo(BaseModel):
    mechanism: str  # defeated_vote | died_committee | died_order_paper | ...
    stage: str | None = None
    occurred_on: date | None = None
    attribution_en: str | None = None
    # Link to the recorded division that killed it, when there was one.
    kill_vote_number: str | None = None
    kill_vote_chamber: str | None = None
    kill_vote_session: str | None = None


class DissenterItem(BaseModel):
    """An MP who broke party ranks on one of this bill's recorded votes."""

    person_slug: str
    full_name: str
    image_url: str | None = None
    party_slug: str | None = None
    ballot: str
    vote_number: str
    session: str
    chamber: str


class BillListItem(BaseModel):
    session: str
    chamber: str
    number: str
    title_en: str
    short_title_en: str | None = None
    status_en: str | None = None
    bill_type: str
    introduced_on: date | None = None
    sponsor_slug: str | None = None
    sponsor_name: str | None = None
    is_omnibus: bool = False
    outcome: str = "pending"
    is_law: bool = False
    death: BillDeathInfo | None = None
    # Plain-language "what this bill is" (published AI summary), for lists.
    one_sentence: str | None = None


class BillDetail(BillListItem):
    # Machine-readable LEGISinfo status code, e.g. "HouseInCommittee".
    status_code: str | None = None
    legisinfo_url: str | None = None
    text_url: str | None = None
    # Human-written by the Library of Parliament — always attributed, no AI.
    official_summary_en: str | None = None
    topics: list[str] = []
    analyses: list[AnalysisState] = []
    related_votes: list[VoteListItem] = []
    sector_impacts: list[dict[str, Any]] = []
    omnibus_components: list[dict[str, Any]] = []
    dissenters: list[DissenterItem] = []
    data_gaps: list[DataGap] = []

File: frontend/lib/api/bills.ts
/** Bills, outcomes, deaths, analyses. */
import { fetchApi } from "./client";
import type { PaginatedResponse } from "./client";
import type { VoteListItem } from "./votes";

export type AnalysisState = {
  analysis_type: string;
  status: string;
  confidence_score?: number | null;
  blocked_reason?: string | null;
  citations?: Array<Record<string, unknown>> | null;
  payload?: Record<string, unknown> | null;
};

export type BillDeathInfo = {
  mechanism: string;
  stage?: string | null;
  occurred_on?: string | null;
  attribution_en?: string | null;
  kill_vote_number?: string | null;
  kill_vote_chamber?: string | null;
  kill_vote_session?: string | null;
};

export type BillListItem = {
  session: string;
  chamber: string;
  number: string;
  title_en: string;
  short_title_en?: string | null;
  status_en?: string | null;
  bill_type: string;
  introduced_on?: string | null;
  sponsor_slug?: string | null;
  sponsor_name?: string | null;
  is_omnibus: boolean;
  outcome: string;
  is_law: boolean;
  death?: BillDeathInfo | null;
  one_sentence?: string | null;
};

export type BillDissenter = {
  person_slug: string;
  full_name: string;
  image_url?: string | null;
  party_slug?: string | null;
  ballot: string;
  vote_number: string;
  session: string;
  chamber: string;
};

export type BillDetail = BillListItem & {
  status_code?: string | null;
  legisinfo_url?: string | null;
  text_url?: string | null;
  official_summary_en?: string | null;
  topics: string[];
  analyses: AnalysisState[];
  related_votes: VoteListItem[];
  sector_impacts: Array<Record<string, unknown>>;
  omnibus_components: Array<Record<string, unknown>>;
  dissenters?: BillDissenter[];
  data_gaps: Array<{
    code: string;
    label: string;
    detail: string;
  }>;
};

export function listBills(params?: { outcomeGroup?: string; topic?: string; offset?: string }) {
  const searchParams = new URLSearchParams();
  if (params?.outcomeGroup) searchParams.set("outcome_group", params.outcomeGroup);
  if (params?.topic) searchParams.set("topic", params.topic);
  if (params?.offset) searchParams.set("offset", params.offset);
  const qs = searchParams.toString();
  return fetchApi<PaginatedResponse<BillListItem>>(`/bills${qs ? `?${qs}` : ""}`);
}

export type GraveyardSummary = {
  total: number;
  mechanisms: Array<{ mechanism: string; count: number }>;
};

/**
 * "How bills die" — mechanism counts across every dead bill. The API caps
 * pages at 100, so this walks the pages (each fetch is cached by Next for
 * the same revalidate window). Hard-capped to stay cheap; if the graveyard
 * ever outgrows the cap the summary is computed from what was fetched and
 * still labeled by its real total.
 */
export async function getGraveyardSummary(): Promise<GraveyardSummary | null> {
  const first = await fetchApi<PaginatedResponse<BillListItem>>(`/bills?outcome_group=dead&limit=100`);
  if (!first) return null;
  const total = first.meta.total;
  const pages = Math.min(Math.ceil(total / 100), 12);
  const rest = await Promise.all(
    Array.from({ length: Math.max(0, pages - 1) }, (_, i) =>
      fetchApi<PaginatedResponse<BillListItem>>(`/bills?outcome_group=dead&limit=100&offset=${(i + 1) * 100}`)
    )
  );
  const counts = new Map<string, number>();
  for (const page of [first, ...rest]) {
    for (const bill of page?.items ?? []) {
      const mechanism = bill.death?.mechanism ?? bill.outcome;
      counts.set(mechanism, (counts.get(mechanism) ?? 0) + 1);
    }
  }
  return {
    total,
    mechanisms: [...counts.entries()]
      .map(([mechanism, count]) => ({ mechanism, count }))
      .sort((a, b) => b.count - a.count)
  };
}

export function getBill(session: string, number: string) {
  return fetchApi<BillDetail>(`/bills/${session}/${number}`, { strict: true });
}
